pkg/detector/cleanup.go

```diff
@@ -0,0 +1,196 @@
+func (d *ResourceDetector) CleanupPolicyRefClusterScopeResources(policyName string, crb *workv1alpha2.ClusterResourceBinding) error {
+	labels := ClusterPropagationPolicyRefLabels
+	annotations := ClusterPropagationPolicyRefAnnotations
+
+	// Must remove the labels from the resource template ahead of ClusterResourceBinding, otherwise might lose the chance
+	// to do that in a retry loop(in particular, the label was successfully removed from ClusterResourceBinding, but
+	// resource template not), since the ClusterResourceBinding will not be listed again.
+	if err := d.cleanupObjectRefLabelsAndAnnotations(crb.Spec.Resource, labels, annotations); err != nil {
+		klog.Errorf("Failed to clean up labels and annotations from resource(%s-%s) when cluster propagation policy(%s) removing, error: %v",
+			crb.Spec.Resource.Kind, crb.Spec.Resource.Name, policyName, err)
+		return err
+	}
+
+	// Clean up the labels and annotations from the reference binding so that the karmada scheduler won't reschedule the binding.
+	if err := d.CleanupCRBLabelsAndAnnotations(crb, labels, annotations); err != nil {
+		klog.Errorf("Failed to clean up labels and annotations from resource binding(%s/%s) when cluster propagation policy(%s) removing, error: %v",
+			crb.Namespace, crb.Name, policyName, err)
+		return err
+	}
+	return nil
+}
+
+// removes labels from object referencing by objRef.
+func (d *ResourceDetector) cleanupObjectRefLabelsAndAnnotations(objRef workv1alpha2.ObjectReference, labels, annotations []string) error {
+	workload, err := helper.FetchResourceTemplate(d.DynamicClient, d.InformerManager, d.RESTMapper, objRef)
+	if err != nil {
+		// do nothing if resource template not exist, it might have been removed.
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		klog.Errorf("Failed to fetch resource(kind=%s, %s/%s): %v", objRef.Kind, objRef.Namespace, objRef.Name, err)
+		return err
+	}
+
+	workload = workload.DeepCopy()
+
+	return d.cleanupWorkLoadLabelsAndAnnotations(workload, labels, annotations)
+}
+
+// removes labels and annotations from object.
+func (d *ResourceDetector) cleanupWorkLoadLabelsAndAnnotations(workload *unstructured.Unstructured, labels, annotations []string) error {
+	cleanupMapKeysFunc := func(m map[string]string, keys []string) map[string]string {
+		for _, key := range keys {
+			delete(m, key)
+		}
+		return m
+	}
+
+	workload.SetLabels(cleanupMapKeysFunc(workload.GetLabels(), labels))
+	workload.SetAnnotations(cleanupMapKeysFunc(workload.GetAnnotations(), annotations))
+
+	gvr, err := restmapper.GetGroupVersionResource(d.RESTMapper, workload.GroupVersionKind())
+	if err != nil {
+		klog.Errorf("Failed to delete resource(%s/%s) labels as mapping GVK to GVR failed: %v", workload.GetNamespace(), workload.GetName(), err)
+		return err
+	}
+
+	newWorkload, err := d.DynamicClient.Resource(gvr).Namespace(workload.GetNamespace()).Update(context.TODO(), workload, metav1.UpdateOptions{})
+	if err != nil {
```

Comment:
Just a question: Should we consider the situation where the workload cannot be found here?

Reply:
Yes, we should do nothing if resource template not exist, since it might have been removed. This step has been implemented in 
```go
// removes labels from object referencing by objRef.
func (d *ResourceDetector) cleanupObjectRefLabelsAndAnnotations(objRef workv1alpha2.ObjectReference, labels, annotations []string) error {
	workload, err := helper.FetchResourceTemplate(d.DynamicClient, d.InformerManager, d.RESTMapper, objRef)
	if err != nil {
		// do nothing if resource template not exist, it might have been removed.
		if apierrors.IsNotFound(err) {
			return nil
		}
		klog.Errorf("Failed to fetch resource(kind=%s, %s/%s): %v", objRef.Kind, objRef.Namespace, objRef.Name, err)
		return err
	}

	workload = workload.DeepCopy()

	return d.cleanupWorkLoadLabelsAndAnnotations(workload, labels, annotations)
}
```

Reply:
Clearly, just because you can find this resource here doesn't necessarily mean it will still exist when you update it.

Reply:
Strictly speaking, there is no guarantee that the resource will still exist at the time of update. If the resource no longer exists, an error will be reported during the update and the retry loop will be entered. Then it will be detected that it no longer exists and the loop will exit. There will be one more retry.